src/main/java/com/browserstack/automate/ci/jenkins/BrowserStackReportForBuild.java

```diff
@@ -0,0 +1,194 @@
+            client = new AutomateClient(credentials.getUsername(), credentials.getDecryptedAccesskey());
+        }
+
+        this.browserStackBuild = fetchBrowserStackBuild(client, this.buildName);
+        if (this.browserStackBuild == null) return false;
+
+        this.browserStackSessions = fetchBrowserStackSessions(client, this.browserStackBuild.getId());
+        this.result.addAll(generateSessionsCollection(this.browserStackSessions));
+
+        if (this.result.size() > 0) {
+            this.result.sort(new SessionsSortingComparator());
+            this.resultMeta.addAll(this.result.subList(0, Math.min(this.result.size(), RESULT_META_MAX_SIZE)));
+            return true;
+        }
+
+        return false;
+    }
+
+    private static Build fetchBrowserStackBuild(@Nonnull BrowserStackClient client, @Nonnull String buildName) {
+        Build build = null;
+        try {
+            build = client.getBuildByName(buildName);
+        } catch (BuildNotFound bnfException) {
+            log(logger, "No build found by name: " + buildName);
+        } catch (BrowserStackException bstackException) {
+            log(logger, "BrowserStackException occurred while fetching build: " + bstackException.toString());
+        }
+
+        return build;
+    }
+
+    private static List<Session> fetchBrowserStackSessions(@Nonnull BrowserStackClient client, @Nonnull String buildId) {
+        List<Session> browserStackSessions = new ArrayList<Session>();
+        try {
+            browserStackSessions.addAll(client.getSessions(buildId));
+        } catch (BuildNotFound bnfException) {
+            log(logger, "No build found while fetching sessions for the buildId: " + buildId);
+        } catch (BrowserStackException bstackException) {
+            log(logger, "BrowserStackException occurred while fetching sessions: " + bstackException.toString());
+        }
+
+        return browserStackSessions;
+    }
+
+    private static List<JSONObject> generateSessionsCollection(List<Session> browserStackSessions) {
+        List<JSONObject> sessionsCollection = new ArrayList<JSONObject>();
+        for (Session session: browserStackSessions) {
+            JSONObject sessionJSON = new JSONObject();
+
+            if (session.getName() == null || session.getName().isEmpty()) {
+                sessionJSON.put("name", session.getId());
+            } else {
+                sessionJSON.put("name", session.getName());
+            }
+
+            if (session.getDevice() == null || session.getDevice().isEmpty()) {
+                sessionJSON.put("browser", session.getBrowser());
+            } else {
+                sessionJSON.put("browser", session.getDevice());
+            }
+            sessionJSON.put("os", session.getOs());
+            sessionJSON.put("osVersion", session.getOsVersion());
+            sessionJSON.put("status", session.getBrowserStackStatus());
+
+            if (session.getBrowserStackStatus().equals(session.getStatus())) {
```

Comment:
Is there any possibility any one of them being null?

Reply:
nope...